x509/revocation/ocsp/ocsp.go

```diff
@@ -0,0 +1,422 @@
+// certificate.
+//
+// Invalid responses and parse failures will result in a ParseError.
+// Error responses will result in a ResponseError.
+func ParseResponseForCert(bytes []byte, cert *x509.Certificate, issuer *x509.Certificate) (*Response, error) {
+	var resp responseASN1
+	rest, err := asn1.Unmarshal(bytes, &resp)
+	if err != nil {
+		fullErr := errors.New("This response is malformed: " + err.Error())
+		return nil, fullErr
+	}
+	if len(rest) > 0 {
+		err = errors.New("trailing data in OCSP response")
+		return nil, err
+	}
+
+	if status := ResponseStatus(resp.Status); status != Success {
+		badResponse := &Response{
+			CertificateStatus: "OCSP Responder Error: " + status.String(),
+		}
+		return badResponse, nil
+	}
+
+	if !resp.Response.ResponseType.Equal(idPKIXOCSPBasic) {
+		err = errors.New("bad OCSP response type")
+		return nil, err
+	}
+
+	var basicResp basicResponse
+	_, err = asn1.Unmarshal(resp.Response.Response, &basicResp)
+	if err != nil {
+		return nil, err
+	}
+
+	if n := len(basicResp.TBSResponseData.Responses); n == 0 || cert == nil && n > 1 {
+		err = errors.New("OCSP response contains bad number of responses")
+		return nil, err
+	}
+
+	var singleResp singleResponse
+	if cert == nil {
+		singleResp = basicResp.TBSResponseData.Responses[0]
+	} else {
+		match := false
+		for _, resp := range basicResp.TBSResponseData.Responses {
+			if cert.SerialNumber.Cmp(resp.CertID.SerialNumber) == 0 {
+				singleResp = resp
+				match = true
+				break
+			}
+		}
+		if !match {
+			err = errors.New("no response matching the supplied certificate")
+			return nil, err
+		}
+	}
+
+	ret := &Response{
+		TBSResponseData:    basicResp.TBSResponseData.Raw,
+		Signature:          basicResp.Signature.RightAlign(),
+		SignatureAlgorithm: x509.GetSignatureAlgorithmFromAI(basicResp.SignatureAlgorithm),
+		Extensions:         singleResp.SingleExtensions,
+		SerialNumber:       singleResp.CertID.SerialNumber.String(),
+		ProducedAt:         basicResp.TBSResponseData.ProducedAt,
+		ThisUpdate:         singleResp.ThisUpdate,
+		NextUpdate:         singleResp.NextUpdate,
+	}
+
+	// Handle the ResponderID CHOICE tag. ResponderID can be flattened into
+	// TBSResponseData once https://go-review.googlesource.com/34503 has been
+	// released.
+	rawResponderID := basicResp.TBSResponseData.RawResponderID
+	switch rawResponderID.Tag {
+	case 1: // Name
+		var rdn pkix.RDNSequence
+		if rest, err = asn1.Unmarshal(rawResponderID.Bytes, &rdn); err != nil || len(rest) != 0 {
+			err = errors.New("invalid responder name")
+			return nil, err
+		}
+		ret.RawResponderName = rawResponderID.Bytes
+	case 2: // KeyHash
+		if rest, err = asn1.Unmarshal(rawResponderID.Bytes, &ret.ResponderKeyHash); err != nil || len(rest) != 0 {
+			err = errors.New("invalid responder key hash")
+			return nil, err
+		}
+	default:
+		err = errors.New("invalid responder id tag")
+		return nil, err
+	}
+
+	if len(basicResp.Certificates) > 0 {
```

Comment:
IIRC the CA and the ocsp signer either need to be the same certificate, or siblings...am I reading this wrong, or does this miss that check?

Reply:
I believe that the signer must either be the exact key that signed the cert, or a key that was explicitly delegated to sign the response (and is provided with the response) https://tools.ietf.org/html/rfc6960#section-4.2.2.2. 

I moved this signature checking code into its own function and added commenting to make it more clear https://github.com/zmap/zcrypto/pull/142/commits/e02c563ed732d505ac262c522861302ae333a5be. I don't think that it misses any checks, how does it look now?

Reply:
The delegate at least needs the `id-kp-OCSPSigning` EKU, right?

The only other potential issue I see is if `BasicResponse.Certs[0]` is not the delegate -- i.e. if it is just an intermediate for validating the signer, for validating a delegate in a later index, or totally unrelated. If the spec mandates that the certificate, when present, must be the delegate, then a comment calling that out would be helpful. If not, I don't see a reasonable way to pick one short of scanning all of `Certs` for candidates and then checking the signature.

Reply:
Oh good catch, the delegate definitely needs proper EKU designation. 

From the RFC: "This [delegate] certificate MUST be issued directly to the responder by the cognizant CA" (https://tools.ietf.org/html/rfc6960#section-2.6). I'm interpreting that as "no intermediates allowed," do you think that's correct? Or could that just mean that the delegate chain must link back to the issuing CA?